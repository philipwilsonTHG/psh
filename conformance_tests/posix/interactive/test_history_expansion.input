# Test history expansion functionality
# PSH supports bash-compatible history expansion (v0.33.0 feature)

# Note: History expansion testing is limited in script mode
# This test focuses on the expansion mechanisms that can be tested

# Test basic history expansion patterns (simulated)
echo "Testing history expansion patterns:"

# Set up some command history simulation
echo "Setting up history context..."
true  # command 1
echo "test command"  # command 2
false # command 3

# Test !! pattern (would expand to previous command)
echo "Testing !! expansion:"
echo "Previous command would be: !!"

# Test !n pattern (would expand to command n)
echo "Testing !n expansion:"
echo "Command 1 would be: !1"
echo "Command 2 would be: !2"

# Test !-n pattern (would expand to command n back)
echo "Testing !-n expansion:"
echo "Last command would be: !-1"
echo "Two commands back would be: !-2"

# Test !string pattern (would expand to last command starting with string)
echo "Testing !string expansion:"
echo "Last echo would be: !echo"
echo "Last true would be: !true"

# Test !?string? pattern (would expand to last command containing string)
echo "Testing !?string? expansion:"
echo "Command containing 'test': !?test?"
echo "Command containing 'command': !?command?"

# Test history expansion in different contexts
echo "Testing history expansion contexts:"

# In variable assignment
echo "In variable assignment:"
result="Previous command: !!"
echo "$result"

# In command substitution
echo "In command substitution:"
output=$(echo "Would expand: !!")
echo "$output"

# In conditional
echo "In conditional:"
if echo "History test: !!" > /dev/null; then
    echo "History expansion in conditional"
fi

# Test quote handling with history expansion
echo "Testing quotes with history expansion:"
echo 'Single quotes prevent: !!'
echo "Double quotes allow: !!"
echo "Escaped: \!\!"

# Test history expansion with special characters
echo "Testing special characters in history:"
echo "Exclamation in string: hello!"
echo "Multiple exclamations: !!!"
echo "Exclamation with space: ! space"

# Test history expansion errors (should be handled gracefully)
echo "Testing history expansion error handling:"
echo "Non-existent command: !nonexistent" 2>/dev/null || echo "Error handled"
echo "Invalid number: !999" 2>/dev/null || echo "Error handled"

# Test history substitution modifiers (advanced)
echo "Testing history modifiers:"
echo "Word selection: !!:1"
echo "All arguments: !!:*"
echo "Last argument: !!:$"

# Test history expansion with glob patterns
echo "Testing history with glob patterns:"
echo "Pattern with history: !echo*"

# Test rapid fire pattern (common use case)
echo "Testing rapid fire commands:"
echo "first command"
echo "second command"
echo "Repeat last: !!"

# Test history expansion in loops
echo "Testing history in loops:"
for i in 1 2; do
    echo "Loop iteration $i: !!"
done

# Test history with redirections
echo "Testing history with redirections:"
echo "Command with redirect: !! > /dev/null"

# Test history event not found scenarios
echo "Testing history event not found:"
echo "Empty pattern: !" 2>/dev/null || echo "Empty pattern handled"

echo "History expansion test completed"