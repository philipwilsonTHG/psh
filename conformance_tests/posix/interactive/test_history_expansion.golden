Testing history expansion patterns:
Setting up history context...
test command
Testing !! expansion:
Previous command would be: !!
Testing !n expansion:
Command 1 would be: !1
Command 2 would be: !2
Testing !-n expansion:
Last command would be: !-1
Two commands back would be: !-2
Testing !string expansion:
Last echo would be: !echo
Last true would be: !true
Testing !?string? expansion:
Command containing 'test': !?test?
Command containing 'command': !?command?
Testing history expansion contexts:
In variable assignment:
Previous command: !!
In command substitution:
Would expand: !!
In conditional:
History expansion in conditional
Testing quotes with history expansion:
Single quotes prevent: !!
Double quotes allow: !!
Escaped: \!\!
Testing special characters in history:
Exclamation in string: hello!
Multiple exclamations: !!!
Exclamation with space: ! space
Testing history expansion error handling:
Non-existent command: !nonexistent
Invalid number: !999
Testing history modifiers:
Word selection: !!:1
All arguments: !!:*
Last argument: !!:$
Testing history with glob patterns:
Pattern with history: !echo*
Testing rapid fire commands:
first command
second command
Repeat last: !!
Testing history in loops:
Loop iteration 1: !!
Loop iteration 2: !!
Testing history with redirections:
Command with redirect: !! > /dev/null
Testing history event not found:
Empty pattern: !
History expansion test completed
